botocore/utils.py

```diff
@@ -260,70 +260,83 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
             proxies=get_environ_proxies(self._url),
         )
 
-    def _get_request(self, url, timeout, num_attempts=1):
+    def retrieve_iam_role_credentials(self):
+        try:
+            role_name = self._get_iam_role()
+            if role_name:
+                credentials = self._get_credentials(role_name)
+                return {
+                    'role_name': role_name,
+                    'access_key': credentials['AccessKeyId'],
+                    'secret_key': credentials['SecretAccessKey'],
+                    'token': credentials['Token'],
+                    'expiry_time': credentials['Expiration'],
+                }
+        except _RetriesExceededError:
+            logger.debug("Max number of attempts exceeded (%s) when "
+                         "attempting to retrieve data from metadata service.",
+                         self._num_attempts)
+        return {}
+
+    def _get_iam_role(self):
+        return self._get_request(
+            url=self._url,
+            num_attempts=self._num_attempts,
+            needs_retry=self._is_non_ok_response
+        ).text
+
+    def _get_credentials(self, role_name):
+        r = self._get_request(
+            url=self._url + role_name,
+            num_attempts=self._num_attempts,
+            needs_retry=self._does_not_contain_credentials
+        )
+        return json.loads(r.text)
+
+    def _get_request(self, url, num_attempts=1, needs_retry=None):
         if self._disabled:
             logger.debug("Access to EC2 metadata has been disabled.")
             raise _RetriesExceededError()
-
         headers = {}
         if self._user_agent is not None:
             headers['User-Agent'] = self._user_agent
 
         for i in range(num_attempts):
             try:
-                AWSRequest = botocore.awsrequest.AWSRequest
-                request = AWSRequest(method='GET', url=url, headers=headers)
+                request = botocore.awsrequest.AWSRequest(
+                    method='GET', url=url, headers=headers)
                 response = self._session.send(request.prepare())
+                if needs_retry is not None and needs_retry(response):
+                    logger.debug(
+                        "Metadata service returned retryable response with "
+                        "status code of %s for url: %s, content body: %s",
+                        response.status_code, url, response.content
+                    )
+                else:
+                    return response
             except RETRYABLE_HTTP_ERRORS as e:
                 logger.debug("Caught exception while trying to retrieve "
```

Comment:
While we're at it, might as well update this message to be more clear and consistent with the other log message

Reply:
So something like? 
```python
logger.debug("Caught HTTP exception while making metadata service request to %s: %s', url, e, exc_info=True)
```

Reply:
Toss 'retryable' in there, it's not clear that it was retried